src/main/java/org/whispersystems/textsecuregcm/storage/DirectoryReconciler.java

```diff
@@ -0,0 +1,344 @@
+          throws CertificateException
+  {
+    try {
+      PEMReader       reader      = new PEMReader(new InputStreamReader(new ByteArrayInputStream(pemCaCertificate.getBytes())));
+      X509Certificate certificate = (X509Certificate) reader.readObject();
+
+      if (certificate == null) {
+        throw new CertificateException("No certificate found in parsing!");
+      }
+
+      KeyStore keyStore = KeyStore.getInstance(KeyStore.getDefaultType());
+      keyStore.load(null);
+      keyStore.setCertificateEntry("ca", certificate);
+      return keyStore;
+    } catch (IOException | KeyStoreException ex) {
+      throw new CertificateException(ex.toString());
+    } catch (NoSuchAlgorithmException ex) {
+      throw new AssertionError(ex);
+    }
+  }
+
+  @Override
+  public synchronized void start() throws Exception {
+    running  = true;
+    finished = false;
+    new Thread(this).start();
+  }
+
+  @Override
+  public synchronized void stop() throws Exception {
+    running = false;
+    while (!finished) Util.wait(this);
+  }
+
+  private synchronized boolean sleepWhileRunning(long delay) {
+    long start   = System.currentTimeMillis();
+    long elapsed = 0;
+    while (running && elapsed < delay) {
+      try {
+        wait(delay - elapsed);
+      } catch (InterruptedException ex) {
+      }
+      elapsed = System.currentTimeMillis() - start;
+    }
+    return running;
+  }
+
+  private List<String> readChunk(Optional<String> from, int chunkSize) {
+    try (Timer.Context timer = readChunkTimer.time()) {
+      if (from.isPresent()) {
+        return accountsManager.getAllNumbers(from.get(), chunkSize);
+      } else {
+        return accountsManager.getAllNumbers(chunkSize);
+      }
+    }
+  }
+
+  private Optional<Response> sendChunk(Optional<String> fromNumber, List<String> numbers) {
+    String reconcilePath;
+    if (fromNumber.isPresent()) {
+      reconcilePath = String.format("/v1/directory/reconcile/%s", fromNumber);
+    } else {
+      reconcilePath = "/v1/directory/reconcile";
+    }
+
+    try (Timer.Context timer = sendChunkTimer.time()) {
+      Response response = client.target(serverApiUrl)
+                                .path(reconcilePath)
+                                .request(MediaType.APPLICATION_JSON)
+                                .put(Entity.json(new DirectoryReconciliationRequest(numbers)));
+      if (response.getStatus() != 200) {
+        sendChunkErrorMeter.mark();
+        logger.warn("http error " + response.getStatus());
+      }
+      return Optional.of(response);
+    } catch (ProcessingException ex) {
+      sendChunkErrorMeter.mark();
+      logger.warn("request error: ", ex);
+      return Optional.absent();
+    }
+  }
+
+  private void processChunk(Jedis jedis) {
+    Optional<String> fromNumber     = Optional.fromNullable(jedis.get(LAST_NUMBER_KEY));
+    Optional<String> cachedCountStr = Optional.fromNullable(jedis.get(CACHED_COUNT_KEY));
```

Comment:
Please no abbreviations in variable names ever

Reply:
jeff done d96018c8edb5cf768675c4ae77b201ca69268c5e